### PR TITLE
#16967 ClickHouse Engine info added to tables

### LINKED ISSUE #16967: Make it possible to choose a table engine when creating a new table via data transfer
When DBeaver transfers data to a ClickHouse table with the 'create' mapping, it defaults to the Log engine. We need an option that allows choosing a different table engine.

Ideally, we need a generic approach since we might want to add something similar in the future. For example, we might want to choose a table engine in MySQL/MariaDB, create indexes in Cassandra/Keyspaces, etc.

### PR DESCRIPTION
**Pull request by LonwoLonwo**
2022-07-13 17_32_33-DBeaver Ultimate 22 1 2 - contact_in

2022-07-19 13_08_58-Configure metadata structure

2022-07-19 13_09_24-Configure metadata structure

### CODE REVIEW

**Comment by serge-rider** [comment]
Please validate data transfer for Clickhouse, Oracle, PostgreSQL, Derby and MongoDB.
With and without additional table settings.
Simple table export and export from query.
Save and load task settings.

Also please test data import (regression).

**Comment by LonwoLonwo** [comment]
https://github.com/dbeaver/dbeaver-ee/pull/1785 small ee request

**Comment by LonwoLonwo** [comment]
In case no editable fields (shouuld be for MongoDB as example)

2022-07-19 16_37_48-Configure metadata structure

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+    private Object sqlPanel;
+    private DBEPersistAction[] persistActions;
+    private boolean ddlTabNeedRefresh = true;
+
+    private PropertySourceEditable propertySource;
+
+    public ConfigureMetadataStructureDialog(@NotNull DataTransferWizard wizard,
+                                            @NotNull DatabaseConsumerSettings settings,
+                                            @NotNull DatabaseMappingContainer mapping,
+                                            @NotNull DatabaseConsumerPageMapping pageMapping) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java]:
```suggestion
    public ConfigureMetadataStructureDialog(
        @NotNull DataTransferWizard wizard,
        @NotNull DatabaseConsumerSettings settings,
        @NotNull DatabaseMappingContainer mapping,
        @NotNull DatabaseConsumerPageMapping pageMapping
    ) {
```

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+        TabItem columnsMappingTab = new TabItem(configTabs, SWT.NONE);
+        columnsMappingTab.setText(DTUIMessages.columns_mapping_dialog_shell_text);
+        ColumnsMappingDialog columnsMappingDialog = new ColumnsMappingDialog(wizard, settings, mapping);
+        columnsMappingDialog.createControl(configTabs);
+        columnsMappingTab.setData(columnsMappingDialog);
+        Control pageControl = columnsMappingDialog.getControl();
+        columnsMappingTab.setControl(pageControl);
+
+        if (mapping.getMappingType() != DatabaseMappingType.create
+            && mapping.getMappingType() != DatabaseMappingType.recreate) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java]: Perhaps `DatabaseMappingType` deserves a method for checking for these two? There are already several occurrences of the same pattern that involves checking for both `create` and `recreate` variants.

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+            tableObject = mapping.getTarget();
+        }
+        final DBSObjectContainer container = settings.getContainer();
+        if (container != null
+            && (mapping.getMappingType() == DatabaseMappingType.create
+                || mapping.getMappingType() == DatabaseMappingType.recreate)) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java]: See previous comment

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+                    }
+                    monitor.done();
+                });
+            } catch (InvocationTargetException e) {
+                DBWorkbench.getPlatformUI().showError(
+                    DTUIMessages.database_consumer_page_mapping_title_target_table,
+                    DTUIMessages.database_consumer_page_mapping_message_error_generating_target_table,
+                    e);
+            } catch (InterruptedException e) {
+                // do nothing
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java:141]:
```suggestion
            } catch (InterruptedException ignored) {
```

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+
+        TabItem showDDLTab = new TabItem(configTabs, SWT.NONE);
+        showDDLTab.setText(DTMessages.data_transfer_wizard_page_ddl_name);
+        showDDL(showDDLTab);
+
+        configTabs.setSelection(0);
+        configTabs.addSelectionListener(new SelectionAdapter() {
+            @Override
+            public void widgetSelected(SelectionEvent e) {
+                if (ddlTabNeedRefresh) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java:202]: Maybe it worth negating this `if` statement?

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+                    final int selectionIndex = configTabs.getSelectionIndex();
+                    final Control[] tabList = configTabs.getTabList();
+                    if (tabList.length > 0 && selectionIndex == tabList.length - 1) {
+                        // Refresh DDL tab - it is the last
+                        final DBSObjectContainer container = settings.getContainer();
+                        if (container == null) {
+                            return;
+                        }
+                        DBPDataSource dataSource = container.getDataSource();
+                        setNewTextToDDLTab(container, dataSource);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java]:
```suggestion
                        if (container != null) {
                            setNewTextToDDLTab(container, container.getDataSource());
                        }
```

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+            }
+        });
+
+        return composite;
+    }
+
+    private void createCompositeWithMessage(GridData gd, TabItem tablePropertiesTab, String message) {
+        Composite compositeEmpty = new Composite(configTabs, SWT.NONE);
+        compositeEmpty.setLayout(new GridLayout(1, false));
+        compositeEmpty.setLayoutData(gd);
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java:223]: I'm unsure about the point of having two nested placeholder composites. Can we avoid this?

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+                e);
+            return null;
+        } catch (InterruptedException e) {
+            return null;
+        }
+        return ddl[0];
+    }
+
+    private boolean applySchemaChanges(@NotNull DBSObjectContainer targetContainer,
+                                       @NotNull DatabaseMappingContainer mapping) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java:359]:
```suggestion
    private boolean applySchemaChanges(
        @NotNull DBSObjectContainer targetContainer,
        @NotNull DatabaseMappingContainer mapping
    ) {
```

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java [hunk 9]
```diff
@@ -0,0 +1,420 @@
...
+                } finally {
+                    monitor.done();
+                }
+            });
+            return true;
+        } catch (InvocationTargetException e) {
+            DBWorkbench.getPlatformUI().showError(DTUIMessages.database_consumer_page_mapping_error_schema_save_title,
+                DTUIMessages.database_consumer_page_mapping_error_schema_save_info, e.getTargetException());
+        } catch (InterruptedException e) {
+            // ignore
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java:384]:
```suggestion
        } catch (InterruptedException ignored) {
```

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseMappingContainer.java [hunk 10]
```diff
@@ -306,6 +329,10 @@ public void loadSettings(DBRRunnableContext context, Map<String, Object> setting
                 }
             }
         }
+        Object changedProperties = settings.get("changedProperties");
+        if (changedProperties != null) {
+            rawChangedPropertiesMap = (Map<String, Object>) settings.get("changedProperties");
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseMappingContainer.java]: Use `JSONUtils#getObject` instead

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseMappingContainer.java [hunk 11]
```diff
@@ -263,6 +278,14 @@ public void saveSettings(Map<String, Object> settings) {
                 }
             }
         }
+        if (!CommonUtils.isEmpty(changedPropertiesMap)) {
+            Map<String, Object> propertiesMap = new LinkedHashMap<>();
+            settings.put("changedProperties", propertiesMap);
+            for (Map.Entry<DBPPropertyDescriptor, Object> entry : changedPropertiesMap.entrySet()) {
+                Object value = entry.getValue();
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseMappingContainer.java:286]: This variable is used only once, so it worth inlining it

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java [hunk 12]
```diff
@@ -114,7 +117,12 @@ public static void refreshDatabaseMappings(@NotNull DBRProgressMonitor monitor,
         }
     }
 
-    public static DBEPersistAction[] generateTargetTableDDL(DBRProgressMonitor monitor, DBCExecutionContext executionContext, DBSObjectContainer schema, DatabaseMappingContainer containerMapping) throws DBException {
+    public static DBEPersistAction[] generateTargetTableDDL(
+        @NotNull DBRProgressMonitor monitor,
+        @NotNull DBCExecutionContext executionContext,
+        DBSObjectContainer schema,
+        @NotNull DatabaseMappingContainer containerMapping,
+        @Nullable Map<DBPPropertyDescriptor, Object> changedProperties) throws DBException {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java]:
```suggestion
        @Nullable Map<DBPPropertyDescriptor, Object> changedProperties
    ) throws DBException {
```

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java [hunk 13]
```diff
@@ -114,7 +117,12 @@ public static void refreshDatabaseMappings(@NotNull DBRProgressMonitor monitor,
         }
     }
 
-    public static DBEPersistAction[] generateTargetTableDDL(DBRProgressMonitor monitor, DBCExecutionContext executionContext, DBSObjectContainer schema, DatabaseMappingContainer containerMapping) throws DBException {
+    public static DBEPersistAction[] generateTargetTableDDL(
+        @NotNull DBRProgressMonitor monitor,
+        @NotNull DBCExecutionContext executionContext,
+        DBSObjectContainer schema,
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java]:
```suggestion
        @NotNull DBSObjectContainer schema,
```

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java [hunk 14]
```diff
@@ -226,18 +236,39 @@ private static void getTableFullName(@Nullable DBSObjectContainer schema, @NotNu
         }
     }
 
-    private static DBEPersistAction[] generateStructTableDDL(DBRProgressMonitor monitor, DBCExecutionContext executionContext, DBSObjectContainer schema, DatabaseMappingContainer containerMapping) throws DBException {
+    @NotNull
+    private static SQLObjectEditor<DBSEntity, ?> getTableManager(DBERegistry editorsRegistry, Class<? extends DBSObject> tableClass)
+        throws DBException {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java:254]:
```suggestion
    private static SQLObjectEditor<DBSEntity, ?> getTableManager(
        @NotNull DBERegistry editorsRegistry,
        @NotNull Class<? extends DBSObject> tableClass
    ) throws DBException {
```

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java [hunk 15]
```diff
@@ -335,13 +374,48 @@ private static DBEPersistAction[] generateStructTableDDL(DBRProgressMonitor moni
...
 
-            List<DBEPersistAction> actions = DBExecUtils.getActionsListFromCommandContext(monitor, commandContext, executionContext, options, null);
-            return actions.toArray(new DBEPersistAction[0]);
+            actions.addAll(
+                DBExecUtils.getActionsListFromCommandContext(
+                    monitor,
+                    commandContext,
+                    executionContext,
+                    options,
+                    null));
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java:408]:
```suggestion
            actions.addAll(DBExecUtils.getActionsListFromCommandContext(
                monitor,
                commandContext,
                executionContext,
                options,
                null
            ));
```

File: plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java [hunk 16]
```diff
@@ -335,13 +374,48 @@ private static DBEPersistAction[] generateStructTableDDL(DBRProgressMonitor moni
...
             throw new DBException("Can't create or modify target table", e);
         }
     }
 
+    public static void implementPropertyChanges(
+        DBRProgressMonitor monitor,
+        Map<DBPPropertyDescriptor, Object> changedProperties,
+        DBECommandContext commandContext,
+        DatabaseMappingContainer containerMapping,
+        DBSEntity table) {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.data.transfer/src/org/jkiss/dbeaver/tools/transfer/database/DatabaseTransferUtils.java]: The name of this method is kinda akward. Perhaps it should be called `applyPropertyChanges`?

Also, please specify all required annotations.

```suggestion
        DBSEntity table
    ) {
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseDataSource.java [hunk 17]
```diff
@@ -56,6 +66,19 @@ public ClickhouseDataSource(DBRProgressMonitor monitor, DBPDataSourceContainer c
...
     {
         super(monitor, container, metaModel, new ClickhouseSQLDialect());
+        engineCache.getAllObjects(monitor, this);
+    }
+
+    List<ClickhouseTableEngine> getTableEngines() {
+        final List<ClickhouseTableEngine> cachedObjects = engineCache.getCachedObjects();
+        if (!CommonUtils.isEmpty(cachedObjects)) {
+            cachedObjects.sort(DBUtils.nameComparator());
+        }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseDataSource.java]: Please use `AbstractObjectCache#setListOrderComparator` in cache's constructor.

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseDataSource.java [hunk 18]
```diff
@@ -78,4 +101,28 @@ protected DBPDataSourceInfo createDataSourceInfo(DBRProgressMonitor monitor, @No
         this.getContainer().getPreferenceStore().setValue(ModelPreferences.RESULT_SET_MAX_ROWS_USE_SQL, true);
         return info;
     }
+
+    static class TableEnginesCache extends JDBCObjectCache<ClickhouseDataSource, ClickhouseTableEngine> {
+        @NotNull
+        @Override
+        protected JDBCStatement prepareObjectsStatement(
+            @NotNull JDBCSession session,
+            @NotNull ClickhouseDataSource clickhouseDataSource) throws SQLException {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseDataSource.java:111]:
```suggestion
            @NotNull ClickhouseDataSource clickhouseDataSource
        ) throws SQLException {
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseDataSource.java [hunk 19]
```diff
@@ -78,4 +101,28 @@ protected DBPDataSourceInfo createDataSourceInfo(DBRProgressMonitor monitor, @No
...
+            return session.prepareStatement("SELECT name FROM system.table_engines");
+        }
+
+        @Nullable
+        @Override
+        protected ClickhouseTableEngine fetchObject(
+            @NotNull JDBCSession session,
+            @NotNull ClickhouseDataSource clickhouseDataSource,
+            @NotNull JDBCResultSet dbResult) {
+
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseDataSource.java:121]:
```suggestion
            @NotNull JDBCResultSet dbResult
        ) {
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseMetaModel.java [hunk 20]
```diff
@@ -63,6 +65,30 @@ public GenericSchema createSchemaImpl(@NotNull GenericDataSource dataSource, @Nu
...
+    public boolean isSystemSchema(GenericSchema schema) {
+        return schema.getName().equalsIgnoreCase("INFORMATION_SCHEMA") || schema.getName().equals("system");
+    }
+
+    @Override
+    public JDBCStatement prepareTableLoadStatement(
+        @NotNull JDBCSession session,
+        @NotNull GenericStructContainer owner,
+        @Nullable GenericTableBase table,
+        @Nullable String tableName) throws SQLException {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseMetaModel.java:78]:
```suggestion
        @Nullable String tableName
    ) throws SQLException {
```

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseTable.java [hunk 21]
```diff
@@ -51,10 +55,45 @@ public class ClickhouseTable extends GenericTable implements DBPObjectStatistics
...
+            setDefaultEngine();
+        }
+    }
+
+    @Nullable
+    private ClickhouseTableEngine searchEngine(String engineName) {
+        if (CommonUtils.isNotEmpty(engineName)) {
+            return getDataSource().getEngineByName(engineName);
+        }
+        return null;
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseTable.java:82]: Unecessary check. Can be boiled down to just `getDataSource().getEngineByName(engineName)`.

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseTable.java [hunk 22]
```diff
@@ -51,10 +55,45 @@ public class ClickhouseTable extends GenericTable implements DBPObjectStatistics
...
+    }
+
+    private void setDefaultEngine() {
+        final List<ClickhouseTableEngine> tableEngines = getDataSource().getTableEngines();
+        if (!CommonUtils.isEmpty(tableEngines)) {
+            // Log is one of the simplest ClickHouse engines. It doesn't need special engine parameters
+            engine = tableEngines.stream().filter(e -> e.getName().equals("Log")).findFirst().orElse(tableEngines.get(0));
+        }
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseTable.java:90]: Caches support finding an object by its name. Consider using that API intead.

File: plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/properties/PropertySourceAbstract.java [hunk 23]
```diff
@@ -346,6 +347,18 @@ public boolean getEnableFilters() {
...
 
+    public void addChangedProperties(DBPPropertyDescriptor descriptor, Object newValue) {
+        changedPropertiesValues.put(descriptor, newValue);
+    }
+
+    public void setChangedPropertiesMap(Map<DBPPropertyDescriptor, Object> newMap) {
+        this.changedPropertiesValues = newMap;
+    }
+
+    public Map<DBPPropertyDescriptor, Object> getChangedPropertiesValues() {
```
Comment (ShadelessFox) [inline plugins/org.jkiss.dbeaver.model/src/org/jkiss/dbeaver/runtime/properties/PropertySourceAbstract.java]: Please add nullability annotations.

**Review by ShadelessFox — CHANGES_REQUESTED** [review]

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseMetaModel.java (hunk 20 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseMetaModel.java:78]: Mmm. No.

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseTable.java (hunk 22 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseTable.java:90]: But what about `tableEngines.get(0)` part? It looks nicer in one line.

File: plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseMetaModel.java (hunk 20 above)
Comment (ShadelessFox) [reply plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseMetaModel.java:78]: Yes.

Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.ext.clickhouse/src/org/jkiss/dbeaver/ext/clickhouse/model/ClickhouseMetaModel.java:78]: Nope. I do not love this code style. I love compact code.

File: plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java (hunk 7 above)
Comment (LonwoLonwo) [reply plugins/org.jkiss.dbeaver.data.transfer.ui/src/org/jkiss/dbeaver/tools/transfer/ui/pages/database/ConfigureMetadataStructureDialog.java:223]: Nope

**Comment by LonwoLonwo** [comment]
https://clickhouse.com/docs/en/engines/table-engines/#special-engines

**Comment by Matvey16** [comment]
Verified

**Pull request merged**